File: typescript/libs/system/workspace/adapters/src/lib/session/trackers/push-to-database-tracker.ts
import { IObject } from '@allors/system/workspace/domain';
import { DatabaseState } from '../state/database-state';

export class PushToDatabaseTracker {
  created: Set<IObject>;

  changed: Set<DatabaseState>;

  onCreated(strategy: IObject) {
    (this.created ??= new Set<IObject>()).add(strategy);
  }

  onChanged(state: DatabaseState) {
    if (!state.object.strategy.isNew) {
      (this.changed ??= new Set<DatabaseState>()).add(state);
    }
  }

  onDelete(object: IObject) {
    this.created?.delete(object);
  }
}

File: dotnet/System/Workspace/Adapters/Allors.Workspace.Adapters/Workspace/PushToDatabaseTracker.cs
// <copyright file="PushToDatabaseTracker.cs" company="Allors bv">
// Copyright (c) Allors bv. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>
// <summary>
//   Defines the AllorsChangeSetMemory type.
// </summary>

namespace Allors.Workspace.Adapters
{
    using System.Collections.Generic;

    public sealed class PushToDatabaseTracker
    {
        public ISet<Strategy> Created { get; set; }

        public ISet<Strategy> Changed { get; set; }

        public void OnCreated(Strategy strategy) => (this.Created ??= new HashSet<Strategy>()).Add(strategy);

        public void OnChanged(Strategy state)
        {
            if (!state.IsNew)
            {
                (this.Changed ??= new HashSet<Strategy>()).Add(state);
            }
        }
    }
}
